### PR TITLE
Get complex JSON API creation and serialization working

### PR DESCRIPTION
**Pull request by dfellis**
Works for both the js-runtime and AVM now!

### CODE REVIEW

File: bdd/spec/032_json_spec.sh [hunk 1]
```diff
@@ -41,4 +41,44 @@ null"
...
+
+  Describe "complex JSON type construction and printing"
+    before() {
+      sourceToAll "
+        from @std/app import start, print, exit
+        from @std/json import JSON, toString, JSONBase, JSONNode, IsObject, Null, newJSONObject, newJSONArray, addKeyVal, push
+
+        on start {
+          newJSONObject()
+            .addKeyVal('mixed', 'values')
```
Comment (depombo) [inline bdd/spec/032_json_spec.sh:53]: huge fan of this API

File: std/json.ln [hunk 2]
```diff
@@ -532,16 +535,29 @@ export fn toString(j: JSON): string {
...
+        missingChildren = true;
+      }
+    });
+    return missingChildren;
+  }).eachLin(fn (kv: KeyVal<int64, bool>) {
+    nodeList.set(kv.key, false);
+  });
   // Execute the following block over and over again until there are no more nodes to process
-  seqdo(newseq(pow(2, 64).getOr(0)), fn (): bool {
+  seqdo(newseq(pow(2, 63).getOr(0)), fn (): bool {
```
Comment (depombo) [inline std/json.ln:560]: what's this about?

**Review by depombo — APPROVED** [review]

File: std/json.ln (hunk 2 above)
Comment (dfellis) [reply std/json.ln:560]: Oh, the actual 2^64 doesn't fit in 64-bits, so it was hitting the `.getOr(0)` and then never running more than once for me.

Comment (depombo) [reply std/json.ln:560]: wow, that is a fun bug. we might have other occurrences of `pow(2, 64)` in `root.ln` for the tree implementation no?

Comment (dfellis) [reply std/json.ln:560]: You're right, there is one other place this exists, I will patch it.

**Comment by dfellis** [comment]
Hmm. I seem to have broken something in this PR and the test suite is running forever. Digging in now, but this isn't going to be as cut-and-dried as I thought :(

**Review by cdmistman — APPROVED** [review]

**Pull request merged**